src/main/java/org/opensearch/ad/task/ADBatchTaskRunner.java

```diff
@@ -1087,62 +1089,61 @@ private void detectAnomaly(
         List<AnomalyResult> anomalyResults = new ArrayList<>();
 
         long intervalEndTime = pieceStartTime;
+        AnomalyDetector detector = adTask.getDetector();
         for (int i = 0; i < pieceSize && intervalEndTime < dataEndTime; i++) {
             Optional<double[]> dataPoint = dataPoints.containsKey(intervalEndTime) ? dataPoints.get(intervalEndTime) : Optional.empty();
             intervalEndTime = intervalEndTime + interval;
+            Instant pieceDataStartTime = Instant.ofEpochMilli(intervalEndTime - interval);
+            Instant pieceDataEndTime = Instant.ofEpochMilli(intervalEndTime);
 
-            if (dataPoint.isEmpty()) {
+            if (dataPoint.isEmpty() && detector.getImputationOption() == null) {
                 AnomalyResult anomalyResult = new AnomalyResult(
                     adTask.getConfigId(),
                     adTask.getConfigLevelTaskId(),
                     null,
-                    Instant.ofEpochMilli(intervalEndTime - interval),
-                    Instant.ofEpochMilli(intervalEndTime),
+                    pieceDataStartTime,
+                    pieceDataEndTime,
                     executeStartTime,
                     Instant.now(),
                     "No data in current detection window",
                     Optional.ofNullable(adTask.getEntity()),
-                    adTask.getDetector().getUser(),
+                    detector.getUser(),
                     anomalyDetectionIndices.getSchemaVersion(ADIndex.RESULT),
                     adTask.getEntityModelId()
                 );
                 anomalyResults.add(anomalyResult);
             } else {
-                List<FeatureData> featureData = ParseUtils.getFeatureData(dataPoint.get(), adTask.getDetector());
-                // 0 is placeholder for timestamp. In the future, we will add
-                // data time stamp there.
-                AnomalyDescriptor descriptor = trcf.process(dataPoint.get(), intervalEndTime);
-                double score = descriptor.getRCFScore();
-                if (!adTaskCacheManager.isThresholdModelTrained(taskId) && score > 0) {
+                // dataPoint is empty or or dataPoint may have partial results (missing feature value is Double.NaN) or imputation option is
+                // not null
+                double[] toScore = null;
+                if (dataPoint.isEmpty()) {
+                    toScore = new double[detector.getEnabledFeatureIds().size()];
+                    Arrays.fill(toScore, Double.NaN);
```

Comment:
Why are we filling it here with Double.NaN and not the filledValues, for example the fixed value or previous value, is this done elsewhere.

Reply:
Double.NaN is used to signal we should put the corresponding indices in the missing value array. RCF will fill in fixed value or previous value according to the missing value array.